src/util/hb_arena.c

```diff
@@ -3,60 +3,152 @@
 #endif
 
 #include "../include/util/hb_arena.h"
+#include "../include/macros.h"
 
+#include <assert.h>
 #include <stdbool.h>
+#include <string.h>
 #include <sys/mman.h>
 
-#define KB(kb) (1024 * kb)
-#define MB(mb) (1024 * KB(mb))
+#define hb_arena_for_each_page(allocator, page)                                                                        \
+  for (hb_arena_page_T* page = (allocator)->head; page != NULL; page = page->next)
 
-bool hb_arena_init(hb_arena_T* allocator, size_t size) {
-  allocator->memory = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+static inline size_t hb_arena_align(size_t size, size_t alignment) {
+  return ((size + (alignment - 1)) / alignment) * alignment;
+}
+
+static inline bool hb_arena_page_is_full(hb_arena_page_T* page, size_t required_size) {
```

Comment:
Naming: `has_capacity` may be more appropriate

Is full indicates that it is a general state, the page is full, period, there's no more space left. But this is not exactly what this function does, it asks whether there is still room for another allocation.